airflow/kubernetes/pod_generator_deprecated.py

```diff
@@ -178,7 +178,7 @@ def __init__(
 
         self.container.command = cmds or []
         self.container.args = args or []
-        self.container.image_pull_policy = image_pull_policy
+        self.container.image_pull_policy = image_pull_policy if image_pull_policy else "IfNotPresent"
```

Comment:
Does it complain if you just don’t set it to None? e.g.

```suggestion
        if image_pull_policy:
            self.container.image_pull_policy = image_pull_policy
```

Reply:
^ if that pattern works, that's probably the best for all of these?

Reply:
Yep it fails.

Reply:
This is what I get when I do it :( 
```
    @image_pull_policy.setter
    def image_pull_policy(self, image_pull_policy):
        """Sets the image_pull_policy of this V1Container.
    
        Image pull policy. One of Always, Never, IfNotPresent. Defaults to Always if :latest tag is specified, or IfNotPresent otherwise. Cannot be updated. More info: https://kubernetes.io/docs/concepts/containers/images#updating-images  Possible enum values:  - `\"Always\"` means that kubelet always attempts to pull the latest image. Container will fail If the pull fails.  - `\"IfNotPresent\"` means that kubelet pulls if the image isn't present on disk. Container will fail if the image isn't present and the pull fails.  - `\"Never\"` means that kubelet never pulls an image, but only uses a local image. Container will fail if the image isn't present  # noqa: E501
    
        :param image_pull_policy: The image_pull_policy of this V1Container.  # noqa: E501
        :type: str
        """
        allowed_values = ["Always", "IfNotPresent", "Never"]  # noqa: E501
        if self.local_vars_configuration.client_side_validation and image_pull_policy not in allowed_values:  # noqa: E501
            raise ValueError(
                "Invalid value for `image_pull_policy` ({0}), must be one of {1}"  # noqa: E501
>               .format(image_pull_policy, allowed_values)
            )
E           ValueError: Invalid value for `image_pull_policy` (None), must be one of ['Always', 'IfNotPresent', 'Never']
```

Reply:
That's really unfortunate. This sure feels like a bug on their side to be honest.

Reply:
I'm experimenting with V1Container, and I can only reproduce this when I set `container.image_pull_policy`. How'd it even get set if it didn't get in the `if`?

Reply:
e.g:

```
>>> from kubernetes import client
>>> container = client.V1Container(name="foo")
>>> # good so far
>>> container.image_pull_policy
>>> # still good
>>> container.image_pull_policy = None
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/kubernetes/client/models/v1_container.py", line 298, in image_pull_policy
    .format(image_pull_policy, allowed_values)
ValueError: Invalid value for `image_pull_policy` (None), must be one of ['Always', 'IfNotPresent', 'Never']
```

Reply:
Did you try  kubernetes 23.3.30 ?

Reply:
Yeah, that's with 23.3.0. Trying with Airflows PodGenerator now.

Reply:
I Just tried it again with this (just to be sure) and it failes in 23.3.0 (it works with 22.*  though):  

```
        if not image_pull_policy:
            self.container.image_pull_policy = image_pull_policy
        self.container.ports = ports or []
        self.container.resources = resources
        self.container.volume_mounts = volume_mounts or []

        # Pod Spec
        self.spec = k8s.V1PodSpec(containers=[])
        self.spec.security_context = security_context
        self.spec.tolerations = tolerations
        if not dnspolicy:
            self.spec.dns_policy = dnspolicy
        self.spec.scheduler_name = schedulername
        self.spec.host_network = hostnetwork
        self.spec.affinity = affinity
        self.spec.service_account_name = service_account_name
        self.spec.init_containers = init_containers
        self.spec.volumes = volumes or []
        self.spec.node_selector = node_selectors
        if not restart_policy:
            self.spec.restart_policy = restart_policy
        self.spec.priority_class_name = priority_class_name
```

Reply:
I'm... confused... then. This works locally for me with 23.3.0

```
--- a/airflow/kubernetes/pod_generator_deprecated.py
+++ b/airflow/kubernetes/pod_generator_deprecated.py
@@ -178,7 +178,8 @@ class PodGenerator:
 
         self.container.command = cmds or []
         self.container.args = args or []
-        self.container.image_pull_policy = image_pull_policy
+        if image_pull_policy:
+            self.container.image_pull_policy = image_pull_policy
         self.container.ports = ports or []
         self.container.resources = resources
         self.container.volume_mounts = volume_mounts or []
@@ -187,7 +188,8 @@ class PodGenerator:
         self.spec = k8s.V1PodSpec(containers=[])
         self.spec.security_context = security_context
         self.spec.tolerations = tolerations
-        self.spec.dns_policy = dnspolicy
+        if dnspolicy:
+            self.spec.dns_policy = dnspolicy
         self.spec.scheduler_name = schedulername
         self.spec.host_network = hostnetwork
         self.spec.affinity = affinity
@@ -195,7 +197,8 @@ class PodGenerator:
         self.spec.init_containers = init_containers
         self.spec.volumes = volumes or []
         self.spec.node_selector = node_selectors
-        self.spec.restart_policy = restart_policy
+        if restart_policy:
+            self.spec.restart_policy = restart_policy
         self.spec.priority_class_name = priority_class_name
 
         self.spec.image_pull_secrets = []
```

Reply:
RIGHT. I think I reached the limit of :facepalm: today :)

```
if not -> if: 
```

Reply:
Ah! Yeah, that'd do it.